### PR TITLE
Fix bug with wrong calculation result in `nn.loss.CrossEntropyLoss`

### PR DESCRIPTION
**Pull request by yghstill**
### Problem Description:

When with `weight` parameter, the calculation result is wrong.
Now This OP combines ``softmax``, ``log``, and ``NLLLoss`` together.
### Instructions
```shell
import paddle
import paddle.fluid as fluid
import numpy as np

input = fluid.data(name='input', shape=[5, 100], dtype='float64')
label = fluid.data(name='label', shape=[5], dtype='int64')
weight = fluid.data(name='weight', shape=[100], dtype='float64')
ce_loss = paddle.nn.loss.CrossEntropyLoss(weight=weight, reduction='mean')
output = ce_loss(input, label)
place = fluid.CPUPlace()
exe = fluid.Executor(place)
exe.run(fluid.default_startup_program())
input_data = np.random.random([5, 100]).astype("float64")
label_data = np.random.randint(0, 100, size=(5)).astype(np.int64)
weight_data = np.random.random([100]).astype("float64")
output = exe.run(fluid.default_main_program(),
                        feed={"input": input_data, "label": label_data,"weight": weight_data},
                        fetch_list=[output],
                        return_numpy=True)
print(output)

# imperative mode
import paddle.fluid.dygraph as dg
with dg.guard(place) as g:
    input = dg.to_variable(input_data)
    label = dg.to_variable(label_data)
    weight = dg.to_variable(weight_data)
    ce_loss = paddle.nn.loss.CrossEntropyLoss(weight=weight, reduction='mean')
    output = ce_loss(input, label)
    print(output.numpy())
```

### Document preview
image
image
image

### CODE REVIEW

File: python/paddle/nn/layer/loss.py [hunk 1]
```diff
@@ -119,26 +123,22 @@ def forward(self, input, label):
...
-        if self.weight is not None:
-            if isinstance(self.weight, fluid.framework.Variable):
-                softmax_out = fluid.layers.elementwise_pow(
-                    softmax_out, self.weight, axis=-1)
-            else:
-                raise ValueError(
-                    "The weight' is not a Variable, please convert to Variable.")
-
-        out = fluid.layers.cross_entropy(softmax_out, label)
+        log_softmax_out = fluid.layers.log(softmax_out)
```
Comment (qingqing01) [inline python/paddle/nn/layer/loss.py]: For LogSoftmax, it's better to reuse https://github.com/PaddlePaddle/Paddle/pull/23258 . Even through, that PR is also combiling `softmax` and `log`.   ( https://github.com/PaddlePaddle/Paddle/pull/23258#pullrequestreview-408609589 )

File: python/paddle/nn/layer/loss.py [hunk 2]
```diff
@@ -62,6 +63,8 @@ class CrossEntropyLoss(fluid.dygraph.Layer):
             If :attr:`size_average` is ``'sum'``, the reduced sum loss is returned.
             If :attr:`reduction` is ``'none'``, the unreduced loss is returned.
             Default is ``'mean'``.
+        ignore_index (int64, optional): Specifies a target value that is ignored
```
Comment (qingqing01) [inline python/paddle/nn/layer/loss.py:68]: Add shape description for `input` and `label` in line 53 and line 55. If shape is more than 2D, also add unit testing for them.

File: python/paddle/fluid/tests/unittests/test_cross_entropy_loss.py [hunk 3]
```diff
@@ -20,21 +20,54 @@
...
+    return exps / np.sum(exps)
+
+
+def cross_entropy_loss_np(input,
+                          label,
+                          weight=None,
+                          reduction='mean',
+                          ignore_index=-100):
+    softmax_out = np.apply_along_axis(stable_softmax, -1, input)
+    log_softmax_out = np.log(softmax_out)
```
Comment (qingqing01) [inline python/paddle/fluid/tests/unittests/test_cross_entropy_loss.py]: def a `log_softmax`

File: python/paddle/nn/layer/loss.py (hunk 1 above)
Comment (yghstill) [reply python/paddle/nn/layer/loss.py]: Thanks, DONE

File: python/paddle/nn/layer/loss.py (hunk 2 above)
Comment (yghstill) [reply python/paddle/nn/layer/loss.py:68]: Thanks, DONE

File: python/paddle/fluid/tests/unittests/test_cross_entropy_loss.py (hunk 3 above)
Comment (yghstill) [reply python/paddle/fluid/tests/unittests/test_cross_entropy_loss.py]: Thanks, DONE

**Review by qingqing01 — APPROVED** [review]
LGTM.
